src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        """
+        A function to set the 'IsClean' flag based on the status from all the
+        pipelines which have worked on the workflow (and have put their markers
+        in the 'PlineMarkers' list)
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        wflow['IsClean'] = self._checkClean(wflow)
+        return wflow
+
+    def archive(self, wflow):
+        """
+        Move the workflow to the proper archived status after checking
+        the full cleanup status
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :param  archStatus: Target status to transition after archival
+        :return wflow:
+        """
+        # NOTE: check allowed status transitions with:
+        #       https://github.com/dmwm/WMCore/blob/5961d2229b1e548e58259c06af154f33bce36c68/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L171
+        if not (wflow['IsClean'] or wflow['ForceArchive']):
+            msg = "Not properly cleaned workflow: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+
+        # Check the available status transitions before we decide the final status
+        targetStatusList = RequestStatus.REQUEST_STATE_TRANSITION.get(wflow['RequestStatus'], [])
+        self.logger.info("targetStatusList: %s", targetStatusList)
+        return wflow
+
+    def getMSOutputTransferInfo(self, wflow):
+        """
+        Fetches the transfer information from the MSOutput REST interface for
+        the given workflow.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        headers = {'Accept': 'application/json'}
+        params = {}
+        url = '%s/data/info?request=%s' % (self.msConfig['msOutputUrl'],
+                                           wflow['RequestName'])
+        mgr = RequestHandler()
+        res = mgr.getdata(url, params=params, headers=headers, ckey=ckey(), cert=cert())
+        data = json.loads(res)['result'][0]
+        TransferInfo = data['transferDoc']
+        if TransferInfo is not None and TransferInfo['TransferStatus'] == 'done': # FIXME 'pennding' -> 'done'
+            wflow['TransferDone'] = True
+        return wflow
+
+    def getRucioRules(self, wflow, gran, rucioAcct):
+        """
+        Queries Rucio and builds the relevant list of blocklevel rules for
+        the given workflow
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  gran:    Data granularity to search for Rucio rules. Possible values:
+                        'block' || 'container'
+        :return:        wflow
+        """
+        currPline = wflow['PlineMarkers'][-1]
+        # Find all the output placement rules created by the agents
+        # NOTE: Check MSOutput 'TransferStatus' first and assure it is 'done'
+        #       only then we proceed with building the respective list of rules
+        if wflow['TransferDone']:
```

Comment:
Isn't rejected and aborted-completed workflows have this always set to its default value (False)? If so, then you know what will happen ;)

Reply:
Ok, I was trying to avoid some extra calls to Rucio, but yes... since we decided that  we will for sure try to clean those workflows too, we cannot avoid calling Rucio every time we meet one of them. So I've put them in a separated section in the `_dispatcher()` logic, and now anybody who calls `getRucioRules()` will receive what one was calling for regardless of the Transfer status in `MSOutput`.

Reply:
I just let them all (`['rejected', 'aborted-completed']`) pass through the pipeline and we get all those errors now: [1]. I am trying to figure out what should be done in this situation. 

[1]
```
2020-11-09 15:52:47,030:DEBUG:lock:_enter_create(): value creation lock <dogpile.cache.region._LockWrapper object at 0x7fa930185950> acquired
2020-11-09 15:52:47,031:DEBUG:lock:_enter_create(): Calling creation function for previously expired value
2020-11-09 15:52:47,031:DEBUG:lock:_enter_create(): Released creation lock
2020-11-09 15:52:47,034:DEBUG:connectionpool:_new_conn(): Starting new HTTP connection (1): cmsrucio-int.cern.ch:80
2020-11-09 15:52:47,087:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FGEN-SIM HTTP/1.1" 200 None
2020-11-09 15:52:47,202:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FLHE HTTP/1.1" 200 0
2020-11-09 15:52:47,231:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep2-DIGI_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FGEN-SIM-RAW HTTP/1.1" 200 0
2020-11-09 15:52:47,260:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep3-RECO_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FAODSIM HTTP/1.1" 200 0
2020-11-09 15:52:47,320:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /dids/cms/%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FGEN-SIM HTTP/1.1" 404 233
2020-11-09 15:52:47,341:ERROR:MSRuleCleaner:_dispatchWflow(): plineAgentBlock: General error from pipeline. Workflow: tivanov_StepChain_Tasks_HG2011_Val_201029_112731_6371. Error: 
<@========== WMException Start ==========@>
Exception Class: WMRucioException
Message: Data identifier not found in Rucio: /DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM. Error: Data identifier not found.
Details: Data identifier 'cms:/DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM' not found
	ModuleName : WMCore.Services.Rucio.Rucio
	MethodName : isContainer
	ClassInstance : None
	FileName : /afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py
	ClassName : None
	LineNumber : 759
	ErrorNr : 0

Traceback: 
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 756, in isContainer
    response = self.cli.get_did(scope=scope, name=didName)

  File "/afs/cern.ch/work/t/tivanov/private/WMCoreDev.d/WMCore.venv/lib/python2.7/site-packages/rucio/client/didclient.py", line 394, in get_did
    raise exc_cls(exc_msg)

<@---------- WMException End ----------@>. 
Will retry again in the next cycle.
Traceback (most recent call last):
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py", line 218, in _dispatchWflow
    pline.run(wflow)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 140, in run
    return reduce(lambda obj, functor: functor(obj), self.funcLine, obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 140, in <lambda>
    return reduce(lambda obj, functor: functor(obj), self.funcLine, obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 72, in __call__
    return self.run(obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 75, in run
    return self.func(obj, *self.args, **self.kwargs)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py", line 393, in getRucioRules
    blocks = self.rucio.getBlocksInContainer(dataCont)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 218, in getBlocksInContainer
    if not self.isContainer(container):
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 759, in isContainer
    raise WMRucioException(msg)
WMRucioException: <@========== WMException Start ==========@>
Exception Class: WMRucioException
Message: Data identifier not found in Rucio: /DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM. Error: Data identifier not found.
Details: Data identifier 'cms:/DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM' not found
	ModuleName : WMCore.Services.Rucio.Rucio
	MethodName : isContainer
	ClassInstance : None
	FileName : /afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py
	ClassName : None
	LineNumber : 759
	ErrorNr : 0

Traceback: 
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 756, in isContainer
    response = self.cli.get_did(scope=scope, name=didName)

  File "/afs/cern.ch/work/t/tivanov/private/WMCoreDev.d/WMCore.venv/lib/python2.7/site-packages/rucio/client/didclient.py", line 394, in get_did
    raise exc_cls(exc_msg)

<@---------- WMException End ----------@>
2020-11-09 15:52:47,362:ERROR:MSRuleCleaner:_dispatchWflow(): plineArchive: Archival Error: MSRuleCleanerArchival: Not properly cleaned workflow: tivanov_StepChain_Tasks_HG2011_Val_201029_112731_6371.  Will retry again in the next cycle.
2020-11-09 15:52:47,363:DEBUG:MSRuleCleaner:_execute(): 
```

Reply:
Can't we catch a `WMRucioException` exception and simply pass it (well, log and pass)? See:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Rucio/Rucio.py#L794

In short, if DID hasn't been found, then we assume no data has been injected into Rucio for that container and there is nothing to be cleaned up.

Reply:
I agree the way out of this is to catch the exception as early as we make the request to Rucio for building these lists of clocks. What I do not like in the current situation (and something I try to work around right now) is this silent Exception metamorphosis that happens in the Rucio module [1], from a very well defined, clear and dedicated exception `DataIdentifierNotFound` which is unambiguously identifying the type of error we get from Rucio, to a more generic `WMRucioException` , in which the only clue that the did was not found is hidden inside the error message. And I do not want to deal with string parsing in order to distinguish exception types.


[1]
https://github.com/dmwm/WMCore/blob/51e3e2f6975b92abbbd9bf7291cf7e8d8eabd2e3/src/python/WMCore/Services/Rucio/Rucio.py#L794

Reply:
The idea with that generic WMCore exception was that clients of the Rucio module would not need to import all the possible exceptions that can happen from rucio.common. But I agree that having only one generic exception isn't ideal anymore.
How about you create another exception like `WMRucioDIDNotFound` and raise that one instead? If you do so, this needs to be communicated to the CRAB team though to make sure they are aware of this change.

Reply:
Thanks @amaltaro , tthat will fit my needs  perfectly. I will communicate the changes with Stefano, once I am ready with the so proposed Exception. 

Hi Stefano, @belforte, including you in this comment here, just for a heads up about an upcoming change.

Reply:
Hi Stefano @belforte, here is the change already prepared to be merged - these are the relevant commit and line [1] we are about to change, that may affect CRAB too. We are changing the more general `WMRucioException` we are rising from the `isContainer()` method in the `Rucio` module with a more specialized one, named `WMRucioDIDNotFoundException`. Here bellow  [2], in order to make your checks a little bit easier, I paste the full call tree and the exception generation/handling points in the code for both the `isContainer()` method and the `WMRucioException` itslef. In short if you export the `Rucio` module from WMcore inside CRAB or any of the methods listed bellow you should expect a change of `Exception` raising behavior. But if you do not catch the `WMRucioException` as  it is anywhere, then we shouldn't break the CRAB code with the current change. Hope this message helps to propagate the information correctly.
   


[1]
https://github.com/dmwm/WMCore/pull/10023/commits/2cda2ed9b64c0956c3fce1b36dba9e972bfa3f54#diff-2c120001070c945d3d9b54bda61f1707d86247a1306b2da392374ed611b62716R766

[2]
```
WMRucioException handling:

   raised from:
      WMCore/Services/Rucio/Rucio.py:
      listDataRules:
         from a general exception handling
      evaluateRSEExpression:
         from a InvalidRSEExpression Rucio exception handling
      isContainer:
         from a DataIdentifierNotFound Rucio exception handling
      createReplicationRule:
         from an AccessDenied Rucio exception handling
      getDataLockedAndAvailable_old:
         from an `if` statement
      getParentContainerRules:
         from an `if` statement
      getDataLockedAndAvailable_old:
         from an `if` statement

   handled at:
      WMComponent/RucioInjector/RucioInjectorPoller.py:_insertContainerRulesT0

Call Graph:
   WMCore/Services/Rucio/Rucio.py:
   isContainer:
      getDataLockedAndAvailable_old:
         TO BE DEPRECATED
      getBlocksInContainer:
         getReplicaInfoForBlocks:
            NOT CALLED
         _getContainerLockedAndAvailable_old:
            TO BE DEPRECATED
         getPileupLockedAndAvailable:
             src/python/WMCore/WMSpec/Steps/Fetchers/PileupFetcher.py:_getDatasetLocation
         getContainerLockedAndAvailable:
            NOT CALLED
      getPileupLockedAndAvailable:
         WMCore/WMSpec/Steps/Fetchers/PileupFetcher.py:_getDatasetLocation
      getDataLockedAndAvailable:
         WMCore/WorkQueue/DataLocationMapper.py:locationsFromRucio
         WMCore/WorkQueue/WorkQueue.py:_blockLocationRucioPhedex
         WMCore/WorkQueue/Policy/Start/StartPolicyInterface.py:getDatasetLocations
```